src/librustc/ty/layout.rs

```diff
@@ -907,14 +907,25 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                         let count = (niche_variants.end().as_u32()
                             - niche_variants.start().as_u32()
                             + 1) as u128;
-                        if let Some((field_index, niche, (niche_start, niche_scalar))) = variants[i]
-                            .iter()
-                            .enumerate()
-                            .filter_map(|(i, &field)| {
-                                let niche = field.largest_niche.as_ref()?;
-                                Some((i, niche, niche.reserve(self, count)?))
+                        let mut niche_size = 0;
+                        if let Some((field_index, niche, niche_start, niche_scalar)) =
+                            variants[i].iter().enumerate().fold(None, |acc, (j, &field)| {
+                                let niche = match &field.largest_niche {
+                                    Some(niche) => niche,
+                                    _ => return acc,
+                                };
+                                let ns = niche.available(dl);
+                                if ns <= niche_size {
+                                    return acc;
+                                }
+                                match niche.reserve(self, count) {
+                                    Some(pair) => {
+                                        niche_size = ns;
+                                        Some((j, niche, pair.0, pair.1))
+                                    }
+                                    None => acc,
+                                }
```

Comment:
You don't need this complication, just have `.max_by_key(|(_, niche, _)| niche.available(dl))` before `.and_then(|(_, niche)| Some((i, niche, niche.reserve(self, count)?)))`.

Reply:
I wanted to do that but that's not exactly the same thing. If it fails for the largest niche, but would have succeeded for a smaller niche, we would end up with None instead of using the smaller niche.
Now, maybe this cannot succeed for a smaller niche and fail for a bigger one so in that case that would be indeed best.

Reply:
> Now, maybe this cannot succeed for a smaller niche and fail for a bigger one so in that case that would be indeed best.

Yeah, that's why we want bigger niches in general: they succeed strictly more often.

Reply:
This is what i've done now.

Reply:
But looking at the code of Niche::reserve, it seems it can still fail if `valid_range_contains` despite being the largest niche while a smaller niche would work.  or is that really unlikely?

Reply:
The calculations are equivalent, I think I even double-checked them using bruteforce for small bitwidths.

`niche.reserve(count)` is checking `niche.available() >= count` but using values it has to compute anyway (i.e. the remaining validity range), instead of actually using `available`.